great_expectations/core/batch.py

```diff
@@ -159,15 +159,17 @@ def __hash__(self) -> int:
             ^ hash(self.data_connector_name)
             ^ hash(self.data_asset_name)
         )
-        if self.definition is not None:
+        if self.partition_definition is not None:
             for key, value in self.partition_definition.items():
                 _result_hash = _result_hash ^ hash(key) ^ hash(str(value))
         return _result_hash
 
 
-class BatchRequest(DictDot):
+class BatchRequestBase(DictDot):
     """
-    This class contains all attributes of a batch_request.
+    This class is for internal inter-object protocol purposes only.
+    As such, it contains all attributes of a batch_request, but does not validate them.
+    See the BatchRequest class, which extends BatchRequestBase and validates the attributes.
```

Comment:
@alexsherstinsky this helped with understanding how these changes fit into the larger "picture", but could you tell me a little bit more about the motivation behind this change? 

- what does `BatchRequestBase` enable the code to do which wasn't possible before? 
- what does it simplify? add functionality to?

Reply:
@Shinnnyshinshin : Thanks for this good question -- as `BatchRequestBase` is the crux of this PR.

`BatchRequestBase` is now used for the internal protocol, which @abegong created during the BatchPod, during which time the very `BatchRequest` was used for this purpose.  However, while very convenient for internal data interchange, the `BatchRequest` was the same object as THE `BatchRequest` that is part of the user API for getting batches and/or validators.  This was the point of the difficulty.  In the user-API-accessible `BatchRequest`, we MUST insist on the triple, consisting of `data_source_name`, `data_connector_name`, and `data_asset_name` to be non-NULL -- none of the three can be NULL.  Whereas for the internal protocol, `BatchRequest` was just used as a flexible bag of attributes, which any fields allowed to be NULL.  Hence, now, `BatchRequestBase` will be the attribute bag for the internal protocol use, whereby NULL values will be allowed, as per the internal needs.  The `BatchRequest` extends `BatchRequestBase` and adds strong validation (as described and more) -- and will now properly validate user input.

Reply:
Thank you for the clear explanation :) It makes a lot more sense. I'd like to have this explanation included somewhere.. do you think a version of the above explanation could be added to the docstrings for `BatchRequest` and `BatchRequestBase`? Otherwise it looks good from my end. 

As a related point, I was wondering what you had in mind for the review process. I'd be happy to hit approve on my end, but can hold off if you'd like to have @jcampbell do the final review and approval.

Reply:
@Shinnnyshinshin -- good idea: I added the docstring with the above explanation (with a better phrasing).

I think that approving it now is safe.  There are no changes to the external API -- in fact, it operates more properly.  Internally, the protocol is also cleaner now, whereby it is easier to understand how the attributes are being passed.

So I think this is safe to approve/merge.  Still, just in case, mentioning @jcampbell here (also /cc @abegong).

Thank you!